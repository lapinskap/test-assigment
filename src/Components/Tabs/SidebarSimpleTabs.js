import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';
import { useHistory, Link } from 'react-router-dom';
import { Card, CardBody } from 'reactstrap';

export default function SidebarSimpleTabs({ tabsConfig, defaultKey }) {
  const history = useHistory();
  const locationHash = history.location.hash.replace('#', '');
  const [activeKey, setActiveKey] = useState(locationHash || defaultKey);
  useEffect(() => () => {
    const { hash } = history.location;
    setActiveKey(hash ? hash.replace('#', '') : defaultKey);
  }, [history.location, defaultKey]);

  const currentKey = activeKey || history.location.hash.replace('#', '');

  const currentTab = tabsConfig.find((item) => item.key === currentKey);
  return (
    <Card>
      <div className="forms-wizard-vertical">
        <div style={{ display: 'flex' }}>
          <div style={{ width: '20%', display: 'flex', flexDirection: 'column' }}>
            <ol className="forms-wizard" style={{ width: '100%' }} data-t1="sidebarTabs">
              {
                  tabsConfig.map(({
                    key, name, icon, header,
                  }) => (header ? (
                    <li
                      className="form-wizard-step-header"
                      key={key || name}
                      data-t1={key}
                    >
                      <span className="title">{name}</span>
                      {icon ? <span className="icon">{React.createElement(icon, { size: 20 })}</span> : null}
                    </li>
                  ) : (
                    <Link key={key || name} to={`#${key}`} data-t1={key}>
                      <li
                        className={`form-wizard-step-${currentKey === key ? 'doing' : 'todo'}`}
                        key={key}
                      >
                        {icon ? <em>{React.createElement(icon)}</em> : null}
                        <span>{name}</span>
                      </li>
                    </Link>
                  )))
                        }
            </ol>
          </div>
          <CardBody style={{ display: 'flex', flexDirection: 'column' }}>
            {currentTab ? currentTab.component : false}
          </CardBody>
        </div>
      </div>
    </Card>
  );
}

SidebarSimpleTabs.propTypes = {
  defaultKey: PropTypes.string.isRequired,
  tabsConfig: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string,
    name: PropTypes.string,
    icon: PropTypes.func,
  })).isRequired,
};
