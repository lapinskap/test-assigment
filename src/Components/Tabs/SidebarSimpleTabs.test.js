/* eslint-disable function-call-argument-newline */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/prop-types */
import React from 'react';
import renderer from 'react-test-renderer';
import SidebarSimpleTabs from './SidebarSimpleTabs';
import '../../tests/setupTests';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    location: {
      hash: 'test',
    },
  }),
  Prompt: ({ message }) => <div>{message}</div>,
  Link: ({ to, children, ...props }) => <a href={to} {...props}>{children}</a>,
}));

describe('<SidebarSimpleTabs />', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  function createNodeMock() {
    return {
      getBoundingClientRect() {
        return {
          width: 100,
        };
      },
    };
  }

  let tabsConfig; let
    defaultKey;

  beforeEach(() => {
    tabsConfig = [
      {
        key: '20',
        name: 'test',
        icon: 's-glyph-5',
      },
    ];
    defaultKey = '';
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with default state', () => {
    const tree = renderer
      .create(<SidebarSimpleTabs tabsConfig={tabsConfig} defaultKey={defaultKey} />, { createNodeMock })
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
