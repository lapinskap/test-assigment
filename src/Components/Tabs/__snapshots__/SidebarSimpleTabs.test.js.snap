// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarSimpleTabs /> renders correctly with default state 1`] = `
<div
  className="card"
>
  <div
    className="forms-wizard-vertical"
  >
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
            "width": "20%",
          }
        }
      >
        <ol
          className="forms-wizard"
          data-t1="sidebarTabs"
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <a
            data-t1="20"
            href="#20"
          >
            <li
              className="form-wizard-step-todo"
            >
              <em>
                <s-glyph-5 />
              </em>
              <span>
                test
              </span>
            </li>
          </a>
        </ol>
      </div>
      <div
        className="card-body"
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
          }
        }
      />
    </div>
  </div>
</div>
`;
